release-23.2: metric: add cidr based routing

andrewbaptist opened the pull request:
Backport 7/7 commits from #130051.
Backport 1/1 commits from #116217.

/cc @cockroachdb/release

---

Backport:
  * 4/4 commits from "metric: add cidr based routing" (#128707)
  * 1/1 commits from "util: deflake cidr.TestRefresh" (#130048)
  * 2/2 commits from "backupccl: break read and write metrics by bucket" (#129059)
  * 1/1 commits from "cloud: use custom HTTP client for gcs" (#116217)

Please see individual PRs for details.

/cc @cockroachdb/release

Release Justification: This is a metrics change required for the billing project.

stevendanna committed:
cloud: use custom HTTP client for gcs

Unfortunately, the WithHTTPClient option overrides all other options
when constructing a GCS client. As a result, it appears we can not
both set credentials options and set an HTTP client with custom
configs via the primary API.

Here, we construct a transport using the SDK which allows us to attach
the relevant credential options to the transport directly before
making the HTTP client.

The downside here is that the SDK's NewTransport's documentation says
that it is not intended for end-user use -- so we may expect breakage
in the future.

Epic: none

Release note: None

Commit:
util: add utility to process CIDR records

This commit adds a new utility which can store and efficiently process a
large number of CIDR records by mapping them to a unique name.

Epic: CRDB-41142

Release note: None

Commit:
sql: restructure sql byte metrics to be add metrics

This commit changes the SQL byte metrics to be broken down by cidr block
of the source.

Epic: CRDB-41142

Release note (ops change): Modifies metrics sql.bytesin and sql.bytesout
to be agg metrics if child metrics are enabled.

Commit:
sql: propagate the cidr.Lookup into ExecutorConfig

This commit constructs the cidr.Lookup into the sql server
ExecutorConfig and evalContext. Additionally this enables the cidr
mapping and adds a new configuration parameter for it.

Epic: CRDB-41142

Release note (ops change): Adds a new configuration parameters
server.cidr_mapping_url which maps IPv4 CIDR blocks to arbitrary tag
names.

Commit:
backupccl: break read and write metrics by bucket

Previously the read and write metrics were aggregate across all
containers or buckets. In some situations it is useful to break the
statistics out to detemine how much data is going to each. This PR
allows breaking the data out for the three main clouds.

Epic: CRDB-41142

Release note: None

Commit:
sql: pass an Inc function to writeBuffer

Previously the writeBuffer required a specific Metric implementation as
part of its paramater. This made it more complicated to change the
metric type that was passed in.

Epic: none

Release note: None

Commit:
util: fix return values for cidr http

Previously the cidr http check did not return true when it completed.
This adds the true return and additionally adds logging for other
failure cases.

Epic: none

Release note: None

Commit:
util: add connection metric tracking utility

Adds a utility to cidrLookup to created a DialContext that is tracked
based on the lookup. This works for any third-party libraries that
expose a way to set the DialContext.

Epic: none

Release note: None

Commit:
util: deflake cidr.TestRefresh

Previously the test was assuming the setting would propagate
syncronously. This could fail under stress and race conditions.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130528)
<!-- Reviewable:end -->

andrewbaptist commented:
Note that I needed to backport #116217 to make this backport to be possible. Also the last commit is quite a bit different than the original since the `NewEarlyBootExternalStorageAccessor` doesn't exist on this branch.

Commit:
util: fix race in cidr startup

Previously if the `server.cidr_mapping_url` was set and a node
restarted, there was a race condition where `SetOnChange` for the
setting could be called before the `Start` was called. This could result
in it blocking while attempting to submit to the channel.

Fixes: #130589

Release note: None

jeffswenson reviewed (approved):
LGTM
